### PR TITLE
[DNM]global sort: new merge step

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 1]
```diff
@@ -271,7 +271,7 @@ func (r *byteReader) reload() error {
 	now := r.concurrentReader.now
 	// in read only false -> true is possible
 	if !now && to {
-		r.logger.Info("switch reader mode", zap.Bool("use concurrent mode", true))
+		// r.logger.Info("switch reader mode", zap.Bool("use concurrent mode", true))
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/byte_reader.go:274]: change to debug?

File: br/pkg/storage/gcs_extra.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+)
+
+const (
+	uploadPartTryCnt   = 3 // if uploading a part fails, we should try 3 times in total.
+	MPUInitiateQuery   = "uploads"
+	MPUPartNumberQuery = "partNumber"
+	MPUUploadIDQuery   = "uploadId"
+
+	defaultChunkSize       = 5 * 1024 * 1024 // 5 MB
+	defaultRetry           = 3               // 建议默认3~5次，不要太大，否则将可能会导致成为僵尸任务。
```
Comment (D3Hunter) [inline br/pkg/storage/gcs_extra.go:26]: translate comments

File: br/pkg/storage/gcs_extra.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+		}
+	}
+	return i, nil
+}
+
+// Close waits for the completion of all transfer and generate the final GCS object.
+func (w *GCSWriter) Close() error {
+	if w.currentWorker != nil {
+		// uploads the last part, which is not a full part
+		w.currentWorker.uploadPartAsync()
```
Comment (D3Hunter) [inline br/pkg/storage/gcs_extra.go:166]: handle err

File: br/pkg/storage/gcs_extra.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+func (uw *uploadWorker) full() bool {
+	return uw.offset >= uw.partSize // if true, this worker is ready to upload the full part
+}
+
+func (uw *uploadWorker) uploadPartAsync() error {
+	if uw.err != nil && uw.tryCnt >= uploadPartTryCnt {
+		return fmt.Errorf("failed to upload part %v too many times, err: %w", uw.index, uw.err)
+	}
+	go func() {
+		err := uw.uploadPart(uw.currentPartNumber, uw.buffer[:uw.offset]) // timeout in few seconds
```
Comment (D3Hunter) [inline br/pkg/storage/gcs_extra.go:305]: if we retry upload after some `Write`, the part-num is changed

**Comment by D3Hunter** [comment]
will review again when https://github.com/pingcap/tidb/pull/48939 merged

File: pkg/disttask/importinto/planner.go [hunk 5]
```diff
@@ -285,13 +285,14 @@ func generateImportSpecs(pCtx planner.PlanCtx, p *LogicalPlan) ([]planner.Pipeli
...
-	})
-	return external.GetMaxOverlappingTotal(stats) <= external.MergeSortOverlapThreshold
+	return false
+	// failpoint.Inject("forceMergeSort", func(val failpoint.Value) {
+	// 	in := val.(string)
+	// 	if in == kvGroup || in == "*" {
+	// 		failpoint.Return(false)
+	// 	}
+	// })
+	// return external.GetMaxOverlappingTotal(stats) <= external.MergeSortOverlapThreshold
```
Comment (tangenta) [inline pkg/disttask/importinto/planner.go:295]: Why change this?

**Pull request closed without merging**